### PR TITLE
wasm_engine: change parse() and instantiate() to return bool

### PR DESCRIPTION
**Pull request by axic**
This is beneficial for wabt and wasm3.

### CODE REVIEW

File: test/bench/bench.cpp [hunk 1]
```diff
@@ -40,6 +40,12 @@ void benchmark_parse(
 {
     const auto engine = create_fn();
 
+    // Pre-run for validation
```
Comment (axic) [inline test/bench/bench.cpp]: Not sure if this is preferred or just a check in the for loop.

Comment (chfast) [reply test/bench/bench.cpp]: I think it is fine for now. Just you can omit {}.

File: test/utils/fizzy_engine.cpp [hunk 2]
```diff
@@ -23,9 +23,10 @@ std::unique_ptr<WasmEngine> create_fizzy_engine()
     return std::make_unique<FizzyEngine>();
 }
 
-void FizzyEngine::parse(bytes_view input)
+bool FizzyEngine::parse(bytes_view input)
 {
     m_instance.module = fizzy::parse(input);
```
Comment (chfast) [inline test/utils/fizzy_engine.cpp]: You should catch exception here and return false then.

File: test/bench/bench.cpp (hunk 1 above)
Comment (axic) [reply test/bench/bench.cpp]: Copied the style from `if (!func_idx)` but I guess because had that 2 lines of message.

Comment (axic) [reply test/bench/bench.cpp]: Changed.

File: test/utils/wasm_engine.hpp [hunk 3]
```diff
@@ -32,10 +33,12 @@ class WasmEngine
     virtual void instantiate() = 0;
 
     /// Returns the memory of the internal instance.
+    /// Can return an empty view if no memory is available (exported).
```
[the remark this answers is not in the document]
Comment (axic) [reply test/utils/wasm_engine.hpp]: Added.

File: test/utils/fizzy_engine.cpp (hunk 2 above)
Comment (axic) [reply test/utils/fizzy_engine.cpp]: Added.

**Comment by axic** [comment]
> Can memory in wasm have size 0?

Yes, it can start with 0 pages and be grown.

File: test/utils/fizzy_engine.cpp [hunk 4]
```diff
@@ -23,9 +23,21 @@ std::unique_ptr<WasmEngine> create_fizzy_engine()
...
-    m_instance.module = fizzy::parse(input);
+    try
+    {
+        m_instance.module = fizzy::parse(input);
+    }
+    catch (const fizzy::parser_error&)
+    {
+        return false;
+    }
+    catch (const fizzy::instantiate_error&)
```
Comment (axic) [inline test/utils/fizzy_engine.cpp]: So this needs to be moved to instantiate. Are we also changing instantiate to return bool? Or should it do an `abort()`?

Comment (gumb0) [reply test/utils/fizzy_engine.cpp]: I think it would be good if it returns `bool`, too

Comment (axic) [reply test/utils/fizzy_engine.cpp]: Added.

**Pull request merged**